TRAF4.py

```diff
@@ -0,0 +1,176 @@
+from Bio import SeqIO
+from Bio.SeqUtils import gc_fraction
+
+
+class NuclAcidnucleotideError(ValueError):
+    """Custom error for NucleicAcidSequence classes"""
+    pass
+
+
+class BiologicalSequence(str):
+    def __init__(self, sequence):
+        self.sequence = sequence
+
+    def check_seq(self):
+        valid_nucleotide_symbols = {'A', 'C', 'G', 'T', 'U'}
+        valid_prot_symbols = {'A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y'}
+        seq_symbols = set(self.sequence.upper())
+        if seq_symbols.issubset(valid_nucleotide_symbols):
+            print('NA sequence')
+        elif seq_symbols.issubset(valid_prot_symbols):
+            print('AA sequence')
+        else:
+            raise ValueError('Incorrect sequence input!')
+        
+    def __str__(self):
+        return self.sequence
+    
+    def __repr__(self):
+        return f'BiologicalSequence("{self.sequence}")'
+
+
+class NucleicAcidSequence(BiologicalSequence):
+    def __init__(self, sequence):
+        self.complement_dict = None
+        super().__init__(sequence)
+    
+    def complement(self):
+        if self.complement_dict == None:
+            raise NotImplementedError('It is a basic NA class. You should implement it for descendant class: DNASequence or RNASequence.')
+        result = type(self)(''.join([self.complement_dict[nuc] for nuc in self.sequence]))
+        return result
```

Comment:
В методе `complement` можно использовать метод `.get` для получения значения из словаря. Это позволит избежать ошибок, если нуклеотид отсутствует в словаре.

```suggestion
    def complement(self):
    if not all(nuc in self.complement_dict for nuc in self.sequence):
        raise NuclAcidnucleotideError('Invalid nucleotide in the sequence')
    result = type(self)(''.join([self.complement_dict.get(nuc, '') for nuc in self.sequence]))
    return result
```